USF-922: Add third-party components to drop-in (#150)
commerce-docs/microsite-commerce-storefront: src/content/docs/dropins/all/extending.mdx

Overloaded term in web development. Everything is a component. It's components all the way down. This is why we need to be specific about what kind of component we are talking about. For example, from top-to-bottom, big-to-small, a **drop-in component** can contain multiple **container components** that can contain multiple **slot components** that can contain multiple **UI components**.

</Vocabulary>

## Add third-party components

> In this how-to, we'll walk through the steps needed to integrate a third-party ratings component on the product-details commerce page. We will add the ratings component under the product's title and the detailed reviews to show below the gallery. However, how you fetch and display these reviews can impact performance, especially CLS. In this guide, we'll show you how to:

Reviewer: A screenshot is needed here.

```suggestion
This tutorial describes the steps needed to integrate a third-party ratings component on the product detail page. We will add the ratings component under the product's title, and the detailed reviews will be displayed below the gallery. However, how you fetch and display these reviews can impact performance, especially CLS.

In this guide, we'll show you how to:
```

What is CLS?

- Configure API keys (Yotpo example)
- Emit and listen for ratings events for the stars under the product title
- Delay load the full reviews to improve page performance

### Prerequisites

<Steps>
1. Yotpo (or other third-party) API key — you’ll need this to fetch rating data.
1. Config Excel — a spreadsheet or config file that stores your API key for easy reference.
1. Event Emitter — a mechanism (like the events object) to emit a ratings event and listen for it in the code.
</Steps>

### Step-by-step

<Tasks>

<Task>
### Set the API Key in Config Excel

> - Open the config Excel file and locate the row or cell dedicated to reviews/ratings integration.

Reviewer: I don't see this line in my configs.xlsx file. I downloaded it a few days ago. If it was added since then, state the exact string that they're looking for.

Author: Added more specific descriptions.

- Configure API keys (Yotpo example)
- Emit and listen for ratings events for the stars under the product title
- Delay load the full reviews to improve page performance

### Prerequisites

<Steps>
1. Yotpo (or other third-party) API key — you’ll need this to fetch rating data.
1. Config Excel — a spreadsheet or config file that stores your API key for easy reference.
1. Event Emitter — a mechanism (like the events object) to emit a ratings event and listen for it in the code.
</Steps>

### Step-by-step

<Tasks>

<Task>
### Set the API Key in Config Excel

- Open the config Excel file and locate the row or cell dedicated to reviews/ratings integration.
- Enter the Yotpo API key (or any other provider key) in the appropriate cell.
- Save your changes. Make sure the system can read from this config file at runtime.

:::tip[API Keys!]

> Keep your API keys safe and out of version control. Treat them like a precious family heirloom from ’80s cinema—guard them with your life.

Reviewer: Let's not be this colloquial. No one's interested in my OG Rubik's Cube. Just the first sentence is fine.

Author: Yeah that was me in a stupor of haze

:::

</Task>

<Task>
### Fetch the data

> After the page loads, your code needs to fetch ratings data from Yotpo (or the relevant third-party). Once the API key is set, you can trigger a call to the Yotpo endpoint.

Reviewer: You can stop saying Yotpo-equivalent.

```suggestion
After the page loads, your code needs to fetch ratings data from Yotpo. Once the API key is set, you can trigger a call to the Yotpo endpoint.
```

Below is a sample fetch call. Adjust it to match your environment:

```js
fetch(`https://api.yotpo.com/v1/widget/<YOUR_API_KEY>/products/<PRODUCT_ID>/reviews`)
  .then(response => response.json())
  .then(data => {
    const { average_score, total_reviews } = data;
    events.emit('eds/pdp/ratings', { average: average_score, total: total_reviews });
  })
  .catch(error => {
    console.error('Error fetching reviews:', error);
  });
```

> 1. Replace `<YOUR_API_KEY>` and `<PRODUCT_ID>` with real values.

Reviewer: No Steps block here?

Author: Done.

2. Adapt the API endpoint to the provider’s docs if not using Yotpo.

</Task>

<Task>

> ### Render the Star Rating (Title Slot)

Reviewer: ```suggestion
### Render the star rating (Title slot)
```

> Star ratings appearing right under the product title can cause unwanted Cumulative Layout Shift (CLS) if not planned carefully. To mitigate CLS, reserve space for the star rating element or render it after the reviews are fetched.

Reviewer: Aha, there's the CLS definition. Move it up.

Listen for the event you just emitted:

```js
events.on('eds/pdp/ratings', ({ average, total }) => {
  // Title slot logic
  const titleSlotElement = document.querySelector('.title-slot');

  // Optionally reserve space for the star rating to avoid CLS
  // e.g., setting a placeholder element or CSS min-height

  // Render stars
  titleSlotElement.innerHTML = `
    <div class="star-rating">
      <span>Average Rating: ${average.toFixed(1)}</span>
      <span>(${total} reviews)</span>
    </div>
  `;
});
```
</Task>

<Task>

> ### Delay Loading Full Reviews (Below the Gallery)

Reviewer: ```suggestion
### Delay loading full reviews (Below the gallery)
```

Overloaded term in web development. Everything is a component. It's components all the way down. This is why we need to be specific about what kind of component we are talking about. For example, from top-to-bottom, big-to-small, a **drop-in component** can contain multiple **container components** that can contain multiple **slot components** that can contain multiple **UI components**.

</Vocabulary>

## Add third-party components

> This tutorial describes the steps needed to integrate a third-party component (the Yotpo ratings & reviews component) within the product detail drop-in. We will add one component instance below the product's title (under the title slot), and a second instance below the product gallery.

Reviewer 2: I would be more clear with the Yotpo integration being just an example. There are multiple 3rd party tools that the client can use for the ratings, and we don't want them to assume that they are required to use the same, or that we specifically recommend it. 
It's simply the case that Bulk happens to use the Yotpo API.

Reviewer 3: +1 to what Daniela said. I would even go further and completely abstract away from Yotpo and add a generic example like `api.rating.service.com`. Otherwise we need to keep the docs up to data with whatever changes Yotpo come up with :)
